cpp/examples/arrow/dataset-parquet-scan-example.cc

```diff
@@ -0,0 +1,136 @@
+
+using arrow::dataset::Dataset;
+using arrow::dataset::DataSource;
+using arrow::dataset::DataSourceDiscovery;
+using arrow::dataset::DataSourceVector;
+using arrow::dataset::Expression;
+using arrow::dataset::FieldExpression;
+using arrow::dataset::FileFormat;
+using arrow::dataset::FileSystemDataSourceDiscovery;
+using arrow::dataset::ParquetFileFormat;
+using arrow::dataset::PartitionScheme;
+using arrow::dataset::Scanner;
+using arrow::dataset::ScannerBuilder;
+using arrow::dataset::ScanTask;
+
+#define ABORT_ON_FAILURE(expr)                     \
+  do {                                             \
+    arrow::Status status_ = (expr);                \
+    if (!status_.ok()) {                           \
+      std::cerr << status_.message() << std::endl; \
+      abort();                                     \
+    }                                              \
+  } while (0);
+
+std::shared_ptr<Dataset> GetDatasetFromPath(FileSystem* fs,
+                                            std::shared_ptr<FileFormat> format,
+                                            std::string path) {
+  // Find all files under `path`
+  Selector s;
+  s.base_dir = path;
+  s.recursive = true;
+
+  std::shared_ptr<DataSourceDiscovery> discovery;
+  ABORT_ON_FAILURE(FileSystemDataSourceDiscovery::Make(fs, s, format, &discovery));
+
+  std::shared_ptr<Schema> inspect_schema;
+  ABORT_ON_FAILURE(discovery->Inspect(&inspect_schema));
+
+  std::shared_ptr<DataSource> source;
+  ABORT_ON_FAILURE(discovery->Finish(&source));
+
+  return std::make_shared<Dataset>(DataSourceVector{source}, inspect_schema);
+}
+
+std::shared_ptr<Scanner> GetScannerFromDataset(std::shared_ptr<Dataset> dataset,
+                                               std::vector<std::string> columns,
+                                               std::shared_ptr<Expression> filter) {
+  std::unique_ptr<ScannerBuilder> scanner_builder;
+  ABORT_ON_FAILURE(dataset->NewScan(&scanner_builder));
+
+  if (!columns.empty()) {
+    ABORT_ON_FAILURE(scanner_builder->Project(columns));
+  }
+
+  if (filter != nullptr) {
+    ABORT_ON_FAILURE(scanner_builder->Filter(filter));
+  }
+
+  std::unique_ptr<Scanner> scanner;
+  ABORT_ON_FAILURE(scanner_builder->Finish(&scanner));
+
+  return scanner;
+}
+
+std::shared_ptr<Table> GetTableFromScanner(std::shared_ptr<Scanner> scanner) {
+  std::shared_ptr<Table> table;
+  ABORT_ON_FAILURE(scanner->ToTable(&table));
+  return table;
+}
+
+int main(int argc, char** argv) {
+  auto fs = std::make_shared<LocalFileSystem>();
+  auto format = std::make_shared<ParquetFileFormat>();
+
+  if (argc != 2) {
+    // Fake success for CI purposes.
+    return EXIT_SUCCESS;
+  }
+
+  auto dataset = GetDatasetFromPath(fs.get(), format, argv[1]);
+
+  // Limit the number of returned columns
+  std::vector<std::string> columns{"pickup_at", "dropoff_at", "total_amount"};
+  // Filter the rows on a predicate
+  auto filter = (FieldExpression("total_amount") > 1000.0f).Copy();
```

Comment:
Nit: I think it looks a bit odd to mix the DSL and explicit class names like this. If you don't want to import the string literal operator (see `filter_test.cc`), maybe:
```suggestion
  auto filter = greater(field_ref("total_amount"), scalar(1000.0f));
```

Reply:
I just didn't know what to import to make it work :)